ls1intum/Artemis
# `Programming exercises`: Improve the template/solution diff view

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


The current implementation of the git-diff view is rather simple and unintuitive. E.g. For each entry, there is one separate ace editor pair, even though it would make sense only to show one pair per file. There were also no context lines, which could make the diff hard to put into context.

### Description

I've completely redone the git-diff view:
- There is no longer a difference between DiffReport and FullDiffReport, as the client now fetches the template and solution repos on its own and uses that to display the diff
- There is now one code editor pair entry per file
- Each files diff can be collapsed
- There are now up to 3 context lines (less if an end of the file was reached)
    - This can be made user configurable in the future
    - Currently, there are infinitely many context lines in between changes (This might be changed in the future)
- Deletions and additions are colored in red/green
- There are now placeholder lines (colored gray), so the lines of each editor always match
- Light and dark mode support

See the screenshots for examples

### Steps for Testing

Prerequisites:
- 1 Tutor/Editor/Instructor/Admin
- 1 Programming Exercise

1. Log in to Artemis
2. Navigate to the programming exercise detail page
3. Adapt the template and solution repos to your liking
4. Scroll down on the detail page
5. Click on the "Show" button in the git-diff section
6. Check that the git-diff is shown correctly according to the repos you set up
7. Repeat from step 3 as often as you want

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| GitDiffReportService.java | 67% | 90% |
| TemplateProgrammingExerciseParticipationRepository.java | n/a | 100% |
| git-diff-file-panel.component.ts.html | 90% | 100% |
| git-diff-file.component.ts.html | 89% | 98% |
| git-diff-report.component.ts.html | 86% | 100% |
| git-diff-line-stat.component.ts.html | 100% | 100% |

### Screenshots


A simple deletion:
<img width="1115" alt="grafik" src="https://user-images.githubusercontent.com/18218285/182513750-3cf282e2-3e49-4498-8256-959d02742eae.png">
A simple addition:
![grafik](https://user-images.githubusercontent.com/18218285/182513875-d885f840-853d-436d-b0cf-e5aa8dc00490.png)
Some changes:
![grafik](https://user-images.githubusercontent.com/18218285/182514005-77db9421-ff18-43b6-ac27-dd9198988de7.png)
Some more changes:
![grafik](https://user-images.githubusercontent.com/18218285/182514075-dd8ce2da-30cb-4cb3-ac52-d175a6608958.png)

Error while fetching the repository contents:
<img width="1166" alt="grafik" src="https://user-images.githubusercontent.com/18218285/185141207-1c275c93-7182-474e-af3c-1508610c0725.png">
